### ISSUE TITLE
partsize of 1024 always fails

### ISSUE THREAD

**Issue by rlpowell** [vsespb/mt-aws-glacier#78]
This happens _every_ time with partsize 1024; 512 is fine.  I wish I could do
something between the two.

```
  rlpowell@ds564> mtglacier upload-file --filename /vz/TO_BACKUP/294122_srv.migrated.tgz --set-rel-filename 294122_srv.migrated.tgz --partsize=1024 --concurrency=1
  MT-AWS-Glacier, Copyright 2012-2014 Victor Efimov http://mt-aws.com/ Version 1.114

  PID 172448 Started worker
  PID 172448 Created an upload_id V-OXMdXOurduhbuVlNJxZDncAT21fVKvEELAp6tvcCBiouee3zubWxa27oBIsJ44Je4Ca9OZX80mNbliu9jw94Ul4YyZ






  Error:
  ===REQUEST:
  PUT http://glacier.us-east-1.amazonaws.com/-/vaults/server-storage/multipart-uploads/V-OXMdXOurduhbuVlNJxZDncAT21fVKvEELAp6tvcCBiouee3zubWxa27oBIsJ44Je4Ca9OZX80mNbliu9jw94Ul4
  YyZ
  Authorization: AWS4-HMAC-SHA256 Credential=***REMOVED***/20140523/us-east-1/glacier/aws4_request, SignedHeaders=content-length;content-range;content-type;host;x-amz-content-s
  ha256;x-amz-date;x-amz-glacier-version;x-amz-sha256-tree-hash, Signature=***REMOVED***
  Host: glacier.us-east-1.amazonaws.com
  User-Agent: mt-aws-glacier/1.114 (http://mt-aws.com/) libwww-perl/5.833
  Content-Length: 1073741824
  Content-Range: bytes 0-1073741823/*
  Content-Type: application/octet-stream
  X-Amz-Content-Sha256: 86084fefdab593cc04acf69ed2834883f24a768d2995b349c6dd470dc525041e
  X-Amz-Date: 20140523T012408Z
  X-Amz-Glacier-Version: 2012-06-01
  X-Amz-Sha256-Tree-Hash: ec6fa43f24661d4570bdf1edc79d71454579b2d4a105adffcb3b0a3a7ae9f43d

  ===RESPONSE:
  HTTP/1.1 403 Forbidden
  Date: Fri, 23 May 2014 02:13:38 GMT
  Content-Length: 312
  Content-Type: application/json
  Client-Date: Fri, 23 May 2014 02:13:38 GMT
  Client-Peer: 205.251.242.31:80
  Client-Response-Num: 1
  X-Amzn-RequestId: OZvsp95thlUYzGQVSM-Zf-56VpM1cHBXqDanVvSm2WNnlRM

  {"message":"The value passed in as x-amz-content-sha256 does not match the computed payload hash. Computed digest: 705c911a603cd1f8558602045e0f610fda9530637ed97f9a071732950ec
  5e22f expected hash: 86084fefdab593cc04acf69ed2834883f24a768d2995b349c6dd470dc525041e","code":"InvalidSignatureException","type":"Client"}

  ERROR (child 172448): Unexpected reply from remote server

  EXIT on SIGCHLD (exit_code 1)
  rlpowell@ds564>
  rlpowell@ds564>
```

**Comment by vsespb** [comment]
Hello. Interesting. I need output of
1. `mtglacier --version`
2. `perl -V`
3. your OS/distro name and version

**Comment by vsespb** [comment]
4 .. and that file's size (in bytes)

**Comment by rlpowell** [comment]
This has happened with files ranging in size from 20GiB to 120GiB

rlpowell@ds564> sudo cat /opt/glaciercli/etc/glacier.cfg
#*************************************************************
#

# THIS FILE IS PUPPET MANAGED

#

# ANY CHANGES YOU MAKE WILL BE DESTROYED WITHIN 24 HOURS

#
#*************************************************************

key=AK[snip]
secret=/NQC1[snip]
region=us-east-1
vault=server-storage
journal=/opt/glaciercli/etc/us-east-1-server-backup.journal
partsize=1024
check-max-file-size=5242880
rlpowell@ds564>

rlpowell@ds564> /opt/glaciercli/mtglacier --version
MT-AWS-Glacier, Copyright 2012-2014 Victor Efimov http://mt-aws.com/ Version 1.114

mt-aws-glacier version: 1.114
Perl Version: 5.010001
AutoLoader      5.68    /usr/share/perl5/AutoLoader.pm
Carp    1.11    /usr/share/perl5/Carp.pm
Carp::Heavy     undef   /usr/share/perl5/Carp/Heavy.pm
Class::Struct   0.63    /usr/share/perl5/Class/Struct.pm
Config  undef   /usr/lib64/perl5/Config.pm
Cwd     3.3     /usr/lib64/perl5/Cwd.pm
Digest::SHA     5.47    /usr/lib64/perl5/Digest/SHA.pm
Digest::base    1.16    /usr/share/perl5/Digest/base.pm
DynaLoader      1.10    /usr/lib64/perl5/DynaLoader.pm
Encode  2.35    /usr/lib64/perl5/Encode.pm
Encode::Alias   2.12    /usr/lib64/perl5/Encode/Alias.pm
Encode::Config  2.05    /usr/lib64/perl5/Encode/Config.pm
Encode::Encoding        2.05    /usr/lib64/perl5/Encode/Encoding.pm
Errno   1.11    /usr/lib64/perl5/Errno.pm
Exporter        5.63    /usr/share/perl5/Exporter.pm
Exporter::Heavy 5.63    /usr/share/perl5/Exporter/Heavy.pm
Fcntl   1.06    /usr/lib64/perl5/Fcntl.pm
File::Basename  2.77    /usr/share/perl5/File/Basename.pm
File::Find      1.14    /usr/share/perl5/File/Find.pm
File::Path      2.08    /usr/share/perl5/File/Path.pm
File::Spec      3.3     /usr/share/perl5/File/Spec.pm
File::Spec::Unix        3.3     /usr/share/perl5/File/Spec/Unix.pm
File::Temp      0.22    /usr/share/perl5/File/Temp.pm
File::stat      1.01    /usr/share/perl5/File/stat.pm
FileHandle      2.02    /usr/share/perl5/FileHandle.pm
FindBin 1.50    /usr/share/perl5/FindBin.pm
Getopt::Long    2.38    /usr/share/perl5/Getopt/Long.pm
HTTP::Date      5.831   /usr/share/perl5/HTTP/Date.pm
HTTP::Headers   5.827   /usr/share/perl5/HTTP/Headers.pm
HTTP::Message   5.832   /usr/share/perl5/HTTP/Message.pm
HTTP::Request   5.827   /usr/share/perl5/HTTP/Request.pm
HTTP::Response  5.824   /usr/share/perl5/HTTP/Response.pm
HTTP::Status    5.817   /usr/share/perl5/HTTP/Status.pm
I18N::Langinfo  0.02    /usr/lib64/perl5/I18N/Langinfo.pm
IO      1.25    /usr/lib64/perl5/IO.pm
IO::File        1.14    /usr/lib64/perl5/IO/File.pm
IO::Handle      1.28    /usr/lib64/perl5/IO/Handle.pm
IO::Pipe        1.13    /usr/lib64/perl5/IO/Pipe.pm
IO::Seekable    1.1     /usr/lib64/perl5/IO/Seekable.pm
IO::Select      1.17    /usr/lib64/perl5/IO/Select.pm
JSON::XS        2.27    /usr/lib64/perl5/JSON/XS.pm
LWP     5.833   /usr/share/perl5/LWP.pm
LWP::MemberMixin        undef   /usr/share/perl5/LWP/MemberMixin.pm
LWP::Protocol   5.829   /usr/share/perl5/LWP/Protocol.pm
LWP::UserAgent  5.833   /usr/share/perl5/LWP/UserAgent.pm
List::Util      1.21    /usr/lib64/perl5/List/Util.pm
MIME::Base64    3.08    /usr/lib64/perl5/MIME/Base64.pm
POSIX   1.17    /usr/lib64/perl5/POSIX.pm
PerlIO::encoding        0.11    /usr/lib64/perl5/PerlIO/encoding.pm
Scalar::Util    1.21    /usr/lib64/perl5/Scalar/Util.pm
SelectSaver     1.02    /usr/share/perl5/SelectSaver.pm
Storable        2.20    /usr/lib64/perl5/Storable.pm
Symbol  1.07    /usr/share/perl5/Symbol.pm
Tie::Hash       1.03    /usr/share/perl5/Tie/Hash.pm
Time::Local     1.1901  /usr/share/perl5/Time/Local.pm
URI     1.40    /usr/share/perl5/URI.pm
URI::Escape     3.29    /usr/share/perl5/URI/Escape.pm
XSLoader        0.10    /usr/lib64/perl5/XSLoader.pm
attributes      0.09    /usr/share/perl5/attributes.pm
base    2.14    /usr/share/perl5/base.pm
bytes   1.03    /usr/share/perl5/bytes.pm
common::sense   3.5     /usr/share/perl5/vendor_perl/common/sense.pm
constant        1.17    /usr/share/perl5/constant.pm
integer 1.00    /usr/share/perl5/integer.pm
lib     0.62    /usr/lib64/perl5/lib.pm
overload        1.07    /usr/share/perl5/overload.pm
re      0.09    /usr/lib64/perl5/re.pm
strict  1.04    /usr/share/perl5/strict.pm
utf8    1.07    /usr/share/perl5/utf8.pm
vars    1.01    /usr/share/perl5/vars.pm
warnings        1.06    /usr/share/perl5/warnings.pm
warnings::register      1.01    /usr/share/perl5/warnings/register.pm
OK DONE
ummary of my perl5 (revision 5 version 10 subversion 1) configuration:

  Platform:
    osname=linux, osvers=2.6.32-358.23.2.el6.x86_64, archname=x86_64-linux-thread-multi
    uname='linux fortythree.fnal.gov 2.6.32-358.23.2.el6.x86_64 #1 smp wed oct 16 11:13:47 cdt 2013 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -DDEBUGGING=-g -Dversion=5.10.1 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl5 -Dsitearch=/usr/local/lib64/perl5 -Dprivlib=/usr/share/perl5 -Darchlib=/usr/lib64/perl5 -Dvendorlib=/usr/share/perl5/vendor_perl -Dvendorarch=/usr/lib64/perl5/vendor_perl -Dinc_version_list=5.10.0 -Darchname=x86_64-linux-thread-multi -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_endhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dscriptdir=/usr/bin -Dusesitecustomize'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.7 20120313 (Red Hat 4.4.7-3)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector'
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.12'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib64/perl5/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic'

Characteristics of this binary (from libperl):
  Compile-time options: MULTIPLICITY PERL_DONT_CREATE_GVSV
                        PERL_IMPLICIT_CONTEXT PERL_MALLOC_WRAP USE_64_BIT_ALL
                        USE_64_BIT_INT USE_ITHREADS USE_LARGE_FILES
                        USE_PERLIO USE_REENTRANT_API USE_SITECUSTOMIZE
  Built under linux
  Compiled at Nov 21 2013 12:43:29
  @INC:
    /usr/local/lib64/perl5
    /usr/local/share/perl5
    /usr/lib64/perl5/vendor_perl
    /usr/share/perl5/vendor_perl
    /usr/lib64/perl5
    /usr/share/perl5
    .
rlpowell@ds564> cat /etc/redhat-release
Scientific Linux release 6.4 (Carbon)
rlpowell@ds564> uname -a
Linux ds564.blueboxgrid.com 2.6.32-042stab061.2 #1 SMP Fri Aug 24 09:07:21 MSK 2012 x86_64 x86_64 x86_64 GNU/Linux
rlpowell@ds564>

**Comment by vsespb** [comment]
Ok, I reproduced and found the problem. Old version of Digest::SHA lib is broken (see changelog for 5.63 https://metacpan.org/changes/distribution/Digest-SHA#L135 ).

I had a check for this

https://github.com/vsespb/mt-aws-glacier/blob/0d4114e5f50447c44aafa1015dcb6488fca99484/lib/App/MtAws/Utils.pm#L379

(which is just throws error asking to upgrade the module)

but apparently I was convinced that only 32bit machines with data chunks >= 512MB (i.e. > 256MB) are affected. Now I reproduced problem on 64bit machine with 1GB chunk (and something like 1.1Gb seem to work).

So workaround for is to upgrade `Digest::SHA` to latest version (via `cpan` or `cpanm`),

but I will fix my wrong bad-module-version detection code (or, probably better, fix my code to workaround the problem and to work with buggy versions).

Thank you for report! I will keep the ticket open until it's fully fixed.
